apollo-router/src/configuration/cors.rs

```diff
@@ -84,90 +154,338 @@ impl Cors {
+                response.headers_mut().insert(
+                    ACCESS_CONTROL_ALLOW_ORIGIN,
+                    http::HeaderValue::from_static("*"),
+                );
+            } else if origin_config.is_some() {
+                // Only set the header if we found a matching origin configuration
+                response
+                    .headers_mut()
+                    .insert(ACCESS_CONTROL_ALLOW_ORIGIN, origin.clone());
+            }
+            // If no matching origin config found, don't set the header (origin will be rejected)
+        }
+
+        // Set Access-Control-Allow-Credentials
+        if allow_credentials {
+            response.headers_mut().insert(
+                ACCESS_CONTROL_ALLOW_CREDENTIALS,
+                http::HeaderValue::from_static("true"),
+            );
+        }
+
+        // Set Access-Control-Allow-Headers
+        if !allow_headers.is_empty() {
+            for header in allow_headers {
+                // Use the header value as-is to match test expectations
+                response.headers_mut().append(
+                    ACCESS_CONTROL_ALLOW_HEADERS,
+                    http::HeaderValue::from_str(header)
+                        .unwrap_or_else(|_| http::HeaderValue::from_static("")),
+                );
+            }
         } else {
-            Ok(cors.allow_origin(cors::AllowOrigin::list(parse_values(
-                &self.origins,
-                "origin",
-            )?)))
+            // If no headers are configured, mirror the client's Access-Control-Request-Headers
+            if let Some(request_headers) = request_headers {
+                if let Ok(headers_str) = request_headers.to_str() {
+                    response.headers_mut().insert(
+                        ACCESS_CONTROL_ALLOW_HEADERS,
+                        http::HeaderValue::from_str(headers_str)
+                            .unwrap_or_else(|_| http::HeaderValue::from_static("")),
+                    );
+                }
+            }
+        }
+
+        // Set Access-Control-Expose-Headers
+        if let Some(headers) = expose_headers {
+            for header in headers {
+                // Use the header value as-is to match test expectations
+                response.headers_mut().append(
+                    ACCESS_CONTROL_EXPOSE_HEADERS,
+                    http::HeaderValue::from_str(header)
+                        .unwrap_or_else(|_| http::HeaderValue::from_static("")),
+                );
+            }
+        }
+
+        // Set Access-Control-Allow-Methods (for preflight requests)
+        if is_preflight {
+            for method in methods {
+                response.headers_mut().append(
+                    ACCESS_CONTROL_ALLOW_METHODS,
+                    http::HeaderValue::from_str(method)
+                        .unwrap_or_else(|_| http::HeaderValue::from_static("")),
+                );
+            }
         }
+
+        // Set Access-Control-Max-Age
```

Comment:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Access-Control-Max-Age indicates this header is only on preflight requests.

Also:
https://github.com/tower-rs/tower-http/blob/35a6cb83242b2004352a8a45f97c0c2cb5475254/tower-http/src/cors/mod.rs#L683